### PR TITLE
Impose limit upon multiplayer room names

### LINKED ISSUE #140: Add room name length restriction
1. Add a check on the server side and send back an error to the client.
2. On the client check for that error and forward it as error to rebo.
3. In rebo handle the error similar to how current errors are handled.

### PR DESCRIPTION
**Pull request by LukeSaward1**
Fixes #140

### CODE REVIEW

File: rtil/src/native/level_state.rs [hunk 1]
```diff
@@ -49,6 +49,31 @@ impl LevelState {
...
+            (*LevelState::get_ptr()).end_seconds = es;
+        }
+    }
+
+    pub fn set_end_partial(eps: f32) {
+        unsafe {
+            (*LevelState::get_ptr()).end_partial_seconds = eps;
+        }
+    }
+
```
Comment (oberien) [inline rtil/src/native/level_state.rs:76]: unrelated to this PR, but I'll let it slide

File: rtil/src/threads/rebo/rebo_init.rs [hunk 2]
```diff
@@ -704,6 +709,10 @@ fn get_level() -> i32 {
 fn set_level(level: i32) {
     LevelState::set_level(level);
 }
+#[rebo::function("Tas::set_end_partial")]
+fn set_end_partial(ep: f32) {
+    LevelState::set_end_partial(ep);
+}
```
Comment (oberien) [inline rtil/src/threads/rebo/rebo_init.rs:715]: If you add all 4 functions to `LevelState`, why only expose one of them?

File: server/src/main.rs [hunk 3]
```diff
@@ -183,48 +183,54 @@ async fn handle_socket(socket: WebSocket, state: Arc<StdMutex<State>>) {
                 let _ = local_sender.send(Response::ServerTime(SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_millis() as u64)).await;
             },
             Request::JoinRoom(room_name, player_name, x, y, z) => {
-                log::info!("Player {player_id:?} ({player_name}) joins room {room_name:?}");
+                if room_name.len() > 128 {
+                    log::error!("Room name is greater than 128 chars.");
+                    local_sender.send(Response::RoomNameTooLong).await;
+                }
```
Comment (oberien) [inline server/src/main.rs:189]: Use a guard-if.

File: server/src/main.rs [hunk 4]
```diff
@@ -183,48 +183,54 @@ async fn handle_socket(socket: WebSocket, state: Arc<StdMutex<State>>) {
                 let _ = local_sender.send(Response::ServerTime(SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_millis() as u64)).await;
             },
             Request::JoinRoom(room_name, player_name, x, y, z) => {
-                log::info!("Player {player_id:?} ({player_name}) joins room {room_name:?}");
+                if room_name.len() > 128 {
+                    log::error!("Room name is greater than 128 chars.");
```
Comment (oberien) [inline server/src/main.rs:187]: It's a `warn` not an `error`. Error usually indicates a problem of the server's implementation. Also please add the player-id, player-name and room-name to the message for better debugging.

File: tool/main.re [hunk 5]
```diff
@@ -97,5 +97,5 @@ loop {
             TickMode::Yield => tick_mode = TickMode::Yield,
         }
     }
-    step_frame(tick_mode);
+    step_frame(Option::None, tick_mode);
```
Comment (oberien) [inline tool/main.re:100]: Rebase error

File: tool/multiplayer.re [hunk 6]
```diff
@@ -1,15 +1,20 @@
...
     Ui::new("Multiplayer:", List::of(
         UiElement::Input(Input {
-            label: Text { text: "Join/Create Room" },
+            label: JOIN_ROOM_LABEL,
             input: "",
             onclick: fn(input: string) {
                 if input.len_utf8() == 0 {
-                    return;
+                    JOIN_ROOM_LABEL.text = "Join/Create Room (Error: empty room name)";
+                }
```
Comment (oberien) [inline tool/multiplayer.re:11]: Use a guard-if.

File: tool/multiplayer.re [hunk 7]
```diff
@@ -340,7 +345,7 @@ fn multiplayer_connect() {
         MULTIPLAYER_STATE.risen_clusters.insert(i, 0);
         i += 1;
     }
-    Tas::connect_to_server(Server::Remote);
+    Tas::connect_to_server(Server::Localhost);
```
Comment (oberien) [inline tool/multiplayer.re:348]: Don't push local testing code :)

File: tool/newgame.re [hunk 8]
```diff
@@ -99,10 +99,13 @@ static NEW_GAME_ALL_BUTTONS_COMPONENT = Component {
     },
     on_level_change: fn(old: int, new: int) {},
     on_reset: fn(old: int, new: int) {
-        Tas::set_level(0);
+        // make All Buttons work together with e.g. NGG
+        let level_state = Tas::get_level_state();
+        if level_state.level == 29 {
+            Tas::set_level(0);
+        }
```
Comment (oberien) [inline tool/newgame.re:106]: Rebase error.

File: tool/newgame.re [hunk 9]
```diff
@@ -99,10 +99,13 @@ static NEW_GAME_ALL_BUTTONS_COMPONENT = Component {
...
     },
     on_platforms_change: fn(old: int, new: int) {},
-    on_buttons_change: fn(old: int, new: int) {},
```
Comment (oberien) [inline tool/newgame.re:105]: ?

**Review by oberien — CHANGES_REQUESTED** [review]

File: rtil/src/threads/rebo/rebo_init.rs (hunk 2 above)
Comment (LukeSaward1) [reply rtil/src/threads/rebo/rebo_init.rs:715]: Good point. Will fix.

File: tool/multiplayer.re (hunk 7 above)
Comment (LukeSaward1) [reply tool/multiplayer.re:348]: oh shit.

File: server/src/main.rs (hunk 3 above)
Comment (LukeSaward1) [reply server/src/main.rs:189]: I believe I cannot do so. Doing so would make the response be `RoomNameTooLong` then immediately change to `ReceiveFailed`. I have tested it and I believe the guard-if makes it not work as intended.

Comment (oberien) [reply server/src/main.rs:189]: Take a look at what the other guard-ifs in that function are doing.

Comment (LukeSaward1) [reply server/src/main.rs:189]: Oh my god I had to use `continue`.

Thanks :)

File: tool/newgame.re (hunk 9 above)
Comment (LukeSaward1) [reply tool/newgame.re:105]: Rebase error I guess. Or I'm stupid as hell.

File: tool/multiplayer.re [hunk 10]
```diff
@@ -471,6 +475,8 @@ fn disconnected(reason: Disconnected) {
         Disconnected::ConnectionRefused => MULTIPLAYER_STATE.connection = Connection::Error("Connection Refused"),
         Disconnected::ReceiveFailed => MULTIPLAYER_STATE.connection = Connection::Error("Receive Failed"),
         Disconnected::LocalTimeOffsetTooManyTries => MULTIPLAYER_STATE.connection = Connection::Error("Connection too unstable; couldn't get local time offset"),
+        Disconnected::RoomNameTooLong => MULTIPLAYER_STATE.connection = Connection::Error("Room name too long"),
     }
+    print(f"MPSC: {MULTIPLAYER_STATE.connection}");
```
Comment (LukeSaward1) [inline tool/multiplayer.re:480]: Don't leave debug prints in code. 😠

File: tool/tas.re [hunk 11]
```diff
@@ -209,4 +209,4 @@ static TAS_COMPONENT = Component {
         TAS_STATE.events.push(InputEvent::MouseMoved(x, y));
     },
     on_component_exit: fn() {},
-};
+};
```
Comment (oberien) [inline tool/tas.re:212]: Empty lines at the end of files are convention. Even if they weren't, or you were adding a missing newline here, if that's the only change to the whole file, don't do it in this PR.

File: tool/misc.re [hunk 12]
```diff
@@ -6,7 +6,6 @@ enum ReplayMenuOp {
 
 fn create_replay_menu(op: ReplayMenuOp) -> Ui{
     let mut recording_name_label = Text { text: "Recording name" };
-
```
Comment (oberien) [inline tool/misc.re:9]: Even if removing this one line makes sense, it is the only change in this file and doesn't have to do anything with this PR.

File: server/src/main.rs [hunk 13]
```diff
@@ -221,10 +226,9 @@ async fn handle_socket(socket: WebSocket, state: Arc<StdMutex<State>>) {
                         }
                     }
                 }
-
```
Comment (oberien) [inline server/src/main.rs:224]: Is removing this line really important and useful?

File: server/src/main.rs [hunk 14]
```diff
@@ -221,10 +226,9 @@ async fn handle_socket(socket: WebSocket, state: Arc<StdMutex<State>>) {
...
                 room.players.write().unwrap().insert(player_id, player);
                 *multiplayer_room.lock().await = Some(room);
-            }
+            },
```
Comment (oberien) [inline server/src/main.rs:231]: This change just clutters the PR. If you had modified lines just above and below, maybe, but not alone like this.

File: rtil/src/threads/rebo/rebo_init.rs [hunk 15]
```diff
@@ -705,6 +713,22 @@ fn get_level() -> i32 {
...
+    LevelState::set_start_partial(sp);
+}
+#[rebo::function("Tas::set_end_seconds")]
+fn set_end_seconds(es: i32) {
+    LevelState::set_end_secs(es);
+}
+#[rebo::function("Tas::set_end_partial")]
+fn set_end_partial(ep: f32) {
+    LevelState::set_end_partial(ep);
+}
```
Comment (oberien) [inline rtil/src/threads/rebo/rebo_init.rs:731]: ~

File: rtil/src/native/level_state.rs (hunk 1 above)
Comment (oberien) [inline rtil/src/native/level_state.rs:76]: Does every letter in a variable or function name cost you money? ;) Just name everything the same, even if its 5 characters longer. That way I don't need to remember which layer of abstraction had which different name for literally the same function or variable.

**Review by oberien — CHANGES_REQUESTED** [review]

File: tool/misc.re (hunk 12 above)
Comment (LukeSaward1) [reply tool/misc.re:9]: True.

File: server/src/main.rs [hunk 16]
```diff
@@ -183,6 +183,11 @@ async fn handle_socket(socket: WebSocket, state: Arc<StdMutex<State>>) {
                 let _ = local_sender.send(Response::ServerTime(SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_millis() as u64)).await;
             },
             Request::JoinRoom(room_name, player_name, x, y, z) => {
+                if room_name.len() > 128 {
+                    log::warn!("Player {player_id:?} tried to join room {room_name:?}, but room name is greater than 128 chars.");
```
Comment (oberien) [inline server/src/main.rs:187]: Please log the player_name as well.

**Review by LukeSaward1 — COMMENTED** [review]
No noticeable mistakes in the code, looks good to me.
EDIT: Except I'm stupid and there actually _are_ mistakes.

**Review by oberien — APPROVED** [review]

**Pull request merged**